### PR TITLE
Closes #6828 Add Ajax for LRC

### LINKED ISSUE #6828: 3.17 - Add AJAX & Database for LRC
# Description
The idea is to implement the data process once the AJAX call is sent by the beacon. As a reminder, we will have a unique call for both feature `Lcp` and `Lrc`. 
It would be better for the [`LrcBeacon`](https://github.com/wp-media/rocket-scripts/issues/7) to be implemented alongside or before this issue for testing purpose. 
For new classes structure, a look can be taken onto `AboveTheFold` classes as they will be also implementing/extending the `PerformanceHints` namespace. 

# Scope a solution
- If not created, a new table needs to be created `wpr_lazy_render_content` 
 - This table will have the following schema: 
```
id                              -  bigint(20)
url                             -  varchar(2000)
is_mobile                 -  boolean (tinyint(1))
below_the_fold	-  longtext
error_message	-  longtext
status			-  varchar(255)
modified		        -  timestamp
last_accessed	        -  timestamp
created_at		-  timestamp
```

- Create a `LazyRenderContent` class within `WP_Rocket\Engine\Optimization\LazyRenderContent\Database\Queries`, `WP_Rocket\Engine\Optimization\LazyRenderContent\Database\Rows`, `WP_Rocket\Engine\Optimization\LazyRenderContent\Database\Schemas` & `WP_Rocket\Engine\Optimization\LazyRenderContent\Database\Tables`.

- Create a `Controller` within `WP_Rocket\Engine\Optimization\LazyRenderContent\AJAX` namespace that will implements `WP_Rocket\Engine\Common\PerformanceHints\AJAX\ControllerInterface` 
- Implement the function `add_data()` and `check_data()` 

# Acceptance Criteria

-   on install/update of the plugin, the new table must be created
-   when running the script on a page without data stored yet:
    -   the script must run the check_data and add_data calls with LCR data (can be seen with logs in PHP and/or in the script)
    -   the data is correctly stored in the DB
    -   must validate that when running on desktop, even if there is data for mobile stored on this URL already.
    -   must be validated when running on a page with already OCI data, but no LCR data yet in DB.
-   when running the script on a page with already data for LCR, and OCI activated:
    -   the script must run the check_data and add_data, but nothing related to LCR must be sent in the add_data call.
    -   no data changes in DB for LCR
-   when running the sript on a page with already data for LCR, and OCI deactivated:
    -   the script must run the check_Data and stop after. No add_data call must be triggered.
    -   no data changes in DB for LCR

### PR DESCRIPTION
**Pull request by Khadreal**
# Description
Decouple #6828, second part of the part. This PR #6872 should be merged first.

Fixes #6828 

### User documentation

Add the controller class to insert the **LCR** data into the database.

### Technical documentation
The implementation is done with a controller using an interface and the methods
- `add_data`
- `check_data`

*Explain how this code works. Diagram & drawings are welcomed.*

## Type of change
- [x] Sub-task.

## New dependencies
*List any new dependencies that are required for this change.*

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Documentation

- [x] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [x] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [x] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.

### CODE REVIEW

File: inc/Engine/Common/PerformanceHints/AJAX/AbstractAJAXController.php
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/AJAX/AbstractAJAXController.php:1]: Why did we abstract this?

File: inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php
Comment (jeawhanlee) [inline inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php:1]: We changed the return for the add_data and check_data [here](https://github.com/wp-media/wp-rocket/pull/6861). I already merged the PR

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Comment by jeawhanlee** [comment]
Some conflicts to resolve

File: inc/Engine/Common/PerformanceHints/AJAX/AbstractAJAXController.php
Comment (Khadreal) [reply inc/Engine/Common/PerformanceHints/AJAX/AbstractAJAXController.php:1]: forgot to use this for above the fold class, the same method is used there too, better to abstract it and use than duplicate it

File: inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php
Comment (Khadreal) [reply inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php:1]: Got it, will update the branch

File: inc/Engine/Optimization/LazyRenderContent/ServiceProvider.php [hunk 1]
```diff
@@ -56,5 +58,17 @@ public function register(): void {
 					$this->getContainer()->get( 'lrc_query' ),
 				]
 			);
+
+		$this->getContainer()->addShared( 'lrc_table', LRCTable::class );
+
+		$this->getContainer()->add( 'lrc_query', LRCQuery::class );
+
+		$this->getContainer()->add( 'atf_ajax_controller', AJAXController::class )
```
Comment (remyperona) [inline inc/Engine/Optimization/LazyRenderContent/ServiceProvider.php]:
```suggestion
		$this->getContainer()->add( 'lrc_ajax_controller', AJAXController::class )
```

File: inc/Engine/Optimization/LazyRenderContent/ServiceProvider.php [hunk 2]
```diff
@@ -23,6 +24,7 @@ class ServiceProvider extends AbstractServiceProvider {
 		'lrc_factory',
 		'lrc_table',
 		'lrc_query',
+		'lrc_controller',
```
Comment (remyperona) [inline inc/Engine/Optimization/LazyRenderContent/ServiceProvider.php]:
```suggestion
		'lrc_ajax_controller',
```

File: inc/Engine/Optimization/LazyRenderContent/Database/Table/LazyRenderContent.php [hunk 3]
```diff
@@ -34,6 +34,7 @@ class LazyRenderContent extends AbstractTable {
 	 */
 	protected $upgrades = [];
 
+
```
Comment (remyperona) [inline inc/Engine/Optimization/LazyRenderContent/Database/Table/LazyRenderContent.php]:
```suggestion
```

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		$below_the_fold = [];
+
+		/**
+		 * Filters the maximum number of LRC hashes being saved into the database.
+		 *
+		 * @param int $max_number Maximum number to allow.
+		 * @param string $url Current page url.
+		 * @param string[]|array $hashes Current list of LRC hashes.
+		 */
+		$max_lrc_hashes_number = (int) apply_filters( 'rocket_lrc_hashes_number', 20, $url, $hashes );
```
Comment (jeawhanlee) [inline inc/Engine/Optimization/LazyRenderContent/AJAX/Controller.php]: I think we should introduce our new filter function here

**Review by jeawhanlee — APPROVED** [review]

**Pull request merged**